### PR TITLE
Host local storage support

### PR DESCRIPTION
**Pull request by torchiaf**
#### PR Checklist
- Is this a multi-tenancy feature/bug?
    - [ ] Yes, the relevant RBAC changes are at:
- Do we need to backport changes to the [old Rancher UI](https://github.com/rancher/u), such as RKE1?
    - [ ] Yes, the relevant PR is at:
- Are backend engineers aware of UI changes?
    - [ ] Yes, the backend owner is:

Related Issue https://github.com/harvester/harvester/issues/6057
Related Issue https://github.com/harvester/harvester/issues/5953

### Screenshot/Video

- Add a LVM disk

image

- Add LVM storage class

image

### CODE REVIEW

**Comment by tserong** [comment]
I've done some initial testing of this, and provisioning and unprovisioning LHv1 and LHv2 disks seems to work, but I've found two things so far that need tweaking:

1) This PR sets a storageclass parameter named `engineVersion` with value `LonghornV1` or `LonghornV2`, but it should instead be `dataEngine` with value `v1` or `v2` (see https://raw.githubusercontent.com/longhorn/longhorn/v1.7.1/examples/v2/storageclass.yaml for an example of a V2 storage class).

2) `spec.fileSystem.provisioned` is deprecated in favour of `spec.provision`, but it seems we're still setting `spec.fileSystem.provisioned` here. If I use this PR to provision a Longhorn V2 disk for example, then check the BD CR, I get this:

```
> kubectl -n longhorn-system get bds/989754e4e66edadfd3390974a1aba3f8 -o yaml 
[...]
spec:
  devPath: /dev/sdb
  fileSystem:
    mountPoint: ""
    provisioned: true
  nodeName: harvester-node-0
  provision: false
  provisioner:
    longhorn:
      engineVersion: LonghornV2
[...]
```

...but it should be:

```
spec:
  devPath: /dev/sdb
  fileSystem:
    mountPoint: ""
  nodeName: harvester-node-0
  provision: true
```

**Comment by torchiaf** [comment]
Thanks @tserong , the code is now updated to handle the `v1/v2` engineVersion in Storage Class page and use the new `spec.provision` field in blockDevice crd in Disks assignement.

**Comment by tserong** [comment]
...oh also, one more thing: After a disk _is_ provisioned, do we need to show what provisioner is used when viewing the disks on that node? Because there's no indication of that currently - you just see the disk like this:

image

@Vicente-Cheng, WDTY? Also, I haven't tested what's shown here in the LVM case.

**Comment by torchiaf** [comment]
@tserong  I've fixed the `dataEngine` parameter, thanks.

For the other points:


- We don't show the 'Force Formatted' checkbox if the disk has already been formatted. This mechanism is unchanged from `master`:
  -  We show up the Force Formatted warning message if any of the new disks is formatted -> `blockDevice.spec.fileSystem.forceFormatted`. Could you verify this value? I can't replicate it in my environment.

image

-  Not sure what was the root cause of provisioning errors, let me know if any API parameter from the UI is wrong.

**Comment by Vicente-Cheng** [comment]
> ...oh also, one more thing: After a disk _is_ provisioned, do we need to show what provisioner is used when viewing the disks on that node? Because there's no indication of that currently - you just see the disk like this:
> 
> image
> 
> @Vicente-Cheng, WDTY? Also, I haven't tested what's shown here in the LVM case.

I checked with the latest changes. It looks like @torchiaf already added the provisioner fields.
截圖 2024-09-25 上午1 22 41

I thought the current display looked good to me. WDYT? @tserong 

>We show up the Force Formatted warning message if any of the new disks is formatted -> blockDevice.spec.fileSystem.forceFormatted. Could you verify this value? I can't replicate it in my environment.

I thought we could keep that, but only for the v1 engine.
For v2 engine, we should not let users select any formatted options.
 
>I also have a problem provisioning Longhorn V1 disks. They seem to add successfully, but shortly thereafter, they aren't mounted and are unschedulable. I suspect this is actually a problem in node-disk-manager. I'll report back once I've figured out what's going on with this.

I will check this... It is not related to UI issue. The CRD fields looks good.

BTW, I checked the whole LVM flow, and everything looks fine. Thanks!

**Comment by tserong** [comment]
Thanks @Vicente-Cheng, @torchiaf! Will re-test this morning.

**Comment by tserong** [comment]
Here's what I see in my environment (latest master without this PR) when I go to add a disk. Note that `blockDevice.spec.fileSystem.forceFormatted` is not set at all:

```
  spec:
    devPath: /dev/sda
    fileSystem:
      mountPoint: ""
    nodeName: harvester-node-0
    provision: false
```

The "force formatted" radio button is present and set to yes, and can't be set to no (this is what I'm used to seeing from previous testing).

image

Here's what I see with this PR applied, and the LHv2 data engine disabled, and LVM unavailable. The "force formatted" radio button is still present:

image

Here's what I see if I enable the LHv2 data engine. I can now switch between provisioners, but the "force formatted" radio button is gone, even if I have Longhorn V1 selected:

image

**Comment by tserong** [comment]
> I checked with the latest changes. It looks like @torchiaf already added the provisioner fields.

Ah! My mistake. If I view the node Detail I _do_ see the provisioner:

image

...but if I switch from Detail to Config, I don't see it:

image

Likewise it's not present when Editing the config:

image

File: pkg/harvester/l10n/en-us.yaml [hunk 1]
```diff
@@ -1103,6 +1123,13 @@ harvester:
         no-options: No available tags, please add in the `Host > Storage` page
       migratable:
         label: Migratable
+      lvmVolumeGroupType:
+        label: Volume Group Type
+      lvmVolumeGroup:
+        label: Volume Group Name
+        no-options: No available Volume Grups, please add in the `Host > Storage` page
```
Comment (a110605) [inline pkg/harvester/l10n/en-us.yaml:1130]: typo `Grups`

**Comment by a110605** [comment]
If the `Path` is always empty for new disk created by LVM provisoner, we can hide that field.

Screenshot 2024-09-25 at 3 17 39 PM

**Comment by a110605** [comment]
Seems we don't clean the previous disk setting when deleting disk in host edit page. 
I need to press save first and edit again to modify the disk provisoner. Is that the limitation from backend ? @Vicente-Cheng 

[remove_disk.webm](https://github.com/user-attachments/assets/ea9e2eaf-a593-4de0-b7e4-6b1c4513890e)

**Comment by torchiaf** [comment]
@tserong You are right, ForceFormatted visibility was wrong. I've also fixed the Provisioner visibility in Config page and pushed CRs @a110605 .

**Comment by tserong** [comment]
Thank you @torchiaf, we're looking pretty good! I can see the ForceFormatted radio button, and the provisioner on the Config page, and I've verified the dataEngine setting is correct.

There's one more thing that we need to add, which is: when creating volumes, right now, the access mode is always set to ReadWriteMany. For LHv2 volumes (i.e. any volume created that uses a LH v2 storage class) we need that to be set to ReadWriteOnce. I'm totally cool though if you'd rather get this PR merged first, then make that change separately. Up to you.

Also, weirdly, when I create new storageclasses (either LH v1 or LH v2), volumes I create using those new storage classes  are showing in the UI as encrypted, even though I never tried to enable encryption. Here's the storage class, and encryption is turned off:

image

```
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  creationTimestamp: "2024-09-26T03:41:17Z"
  name: lhv2
  resourceVersion: "56081"
  uid: 827745db-0cea-4b28-8457-5756fcef7dea
parameters:
  dataEngine: v2
  encrypted: "false"
  migratable: "false"
  numberOfReplicas: "3"
  staleReplicaTimeout: "30"
provisioner: driver.longhorn.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

...but if I go look at the list of volumes, the ones backed by newly created storageclasses all have the little encrypted padlock:

image

Same when viewing a VM that uses those volumes:

image

File: pkg/harvester/edit/harvesterhci.io.storage/provisioners/lvm.driver.harvesterhci.io.vue [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import { HCI } from '../../../types';
+import { NODE } from '@shell/config/types';
+import { LVM_TOPOLOGY_LABEL } from '../index.vue';
+
+const DEFAULT_PARAMETERS = [
+  'type',
+  'vgName'
+];
+
+const DEFAUL_TOPOLOGIES = [{
```
Comment (a110605) [inline pkg/harvester/edit/harvesterhci.io.storage/provisioners/lvm.driver.harvesterhci.io.vue:17]: typo `DEFAUL_TOPOLOGIES`

**Comment by torchiaf** [comment]
@tserong @Vicente-Cheng I've changed the `AccessMode` for LonghornV2 disks and limited the `VolumeBindingMode` to WaitForFirstConsumer in Storage Classes create page.

**Comment by tserong** [comment]
Thanks @torchiaf, I've just tested and the AccessMode change for LHv2 volumes works fine when the new volumes are created on the Volumes page. For new volumes created while creating virtual machines though (Virtual Machines: Create: Volumes: Add Volume), it's still using ReadWriteMany.

**Comment by Vicente-Cheng** [comment]
Thanks @torchiaf, I have tested the latest changes.
Looks the `VolumeBindingMode` works good with `WaitForFirstConsumer` with StorageClass page.

But I tested the volume creation. The `accessMode` is still RWX instead RWO. See below
截圖 2024-09-26 下午4 45 51

Create Vol002 from UI, then check the accessMode
截圖 2024-09-26 下午4 46 04

Could you help to check it again?
Thanks!

**Comment by torchiaf** [comment]
I 've fixed the `accessMode` property in Virtual Machine and Volumes create pages, following this rule:

```
- LonghronV1: RWX
- LonghornV2, Lvm: RWO
```

**Review by tserong — APPROVED** [review]
The LHv2 changes are all behaving correctly in my testing. Thanks for all your work on this @torchiaf!

**Review by Vicente-Cheng — APPROVED** [review]
Tested with LVM part (with addon enable/vg creation/volume creation)
Thanks!

**Comment by a110605** [comment]
Are `Volume Group Name` and `Volume Group Type` required when creating Storage Class ?  cc @Vicente-Cheng 
I can successfully create one even NOT choose any volume group name / type 

Screenshot 2024-09-30 at 8 59 01 PM

**Comment by a110605** [comment]
The warning banner is not added via this PR. 

> Seems we don't clean the previous disk setting when deleting disk in host edit page. I need to press save first and edit again to modify the disk provisoner. Is that the limitation from backend ? @Vicente-Cheng
> 
>  remove_disk.webm

Delete unsaved disk and then add disk back gets previous settings. Do we want to fix this behavior ?

**Comment by a110605** [comment]
When adding new disk, if I didn't choose any `volume group` and got error. I'm unable to modify anything.

Screenshot 2024-09-30 at 9 31 16 PM

**Comment by a110605** [comment]
Does Longhorn V2 volume support resize ? cc @tserong @Vicente-Cheng 
I got the LHv2 volume stuck in resizing status while LHv1 volume is good on resizing.  
If not, we should not allow user expanding volume size regarding from backend webhook or disable in frontend

Screenshot 2024-09-30 at 9 34 50 PM
Screenshot 2024-09-30 at 9 35 02 PM

**Comment by a110605** [comment]
Does LHv2 volume support `Export to Image` ?  @tserong  @Vicente-Cheng 
Seems not, we should hide the expot to image action for LHv2 volume.

Screenshot 2024-09-30 at 9 41 59 PM

**Comment by Vicente-Cheng** [comment]
Hi @a110605,
Thanks for these questions, here are some update:

> Are Volume Group Name and Volume Group Type required when creating Storage Class ? cc @Vicente-Cheng
I can successfully create one even NOT choose any volume group name / type

We already have webhook on v0.1.3. I thought you were using the v0.1.2 on the lab environment.
The latest behavior would block the empty vgName/type/nodeName by webhook like below:
截圖 2024-10-01 上午10 17 56

>Delete unsaved disk and then add disk back gets previous settings. Do we want to fix this behavior ?
As I mentioned before, this behavior is related to UI behavior.
Delete the disk w/o saving would not change any CR, so adding this disk back will get the previous settings.
So, If we would like to delete the disk w/o saving and adding back with an empty provisioner, we need to handle it on the UI side.

>When adding new disk, if I didn't choose any volume group and got error. I'm unable to modify anything.
We can remove and re-add it to handle this user error, but I will also create a webhook to check it.

>Does Longhorn V2 volume support resize ? cc @tserong @Vicente-Cheng
I got the LHv2 volume stuck in resizing status while LHv1 volume is good on resizing.
If not, we should disable from frontend or block from backend webhook.

I think v2 volume does not support resize. (REF: https://github.com/longhorn/longhorn/issues/8022)
So, as you mentioned, we should block this operation. cc @tserong 

>Does LHv2 volume support Export to Image ? @tserong @Vicente-Cheng
Seems not, we should hide the expot to image action for LHv2 volume.

+1, cc @tserong 

>Maybe we should disable LHv2 volume to Take a snapshot. cc @tserong , @Vicente-Cheng
There is no restore action for LHv2 volume snapshot to do restore.

Looks reasonable to me if we cannot restore the v2 volume. cc @tserong

**Comment by tserong** [comment]
Agree with @Vicente-Cheng to disable expand and export to image.  As for snapshots, I have created a LHv2 volume snapshot and (seem to have) successfully restored it to a new volume, so IMO we should leave this enabled.

My one question is: can we do this extra little bit of work separately, or should we do it as part of this PR, given we really want to get this in for RC1?

**Comment by Vicente-Cheng** [comment]
> Agree with @Vicente-Cheng to disable expand and export to image. As for snapshots, I have created a LHv2 volume snapshot and (seem to have) successfully restored it to a new volume, so IMO we should leave this enabled.
> 
> My one question is: can we do this extra little bit of work separately later, or should we do it as part of this PR, bearing in mind we really want to get this PR merged as soon as possible for RC1? (My preference is to merge as-is if there's no other issues)

Yeah, I agree!
I thought the most work looked great. We could fix others after the rc1.
We could get more feedback if we can ship this UI change with RC1.
WDYT? cc @a110605, @torchiaf, @bk201

**Comment by torchiaf** [comment]
Let's open a new Issue/PR to change the remaining functionalities.

**Comment by torchiaf** [comment]
@mergify backport release-harvester-v1.4

**Pull request merged**